Content.Server/PipePainter/PipePainterSystem.cs

```diff
@@ -0,0 +1,138 @@
+using Content.Server.Atmos.Piping.EntitySystems;
+using Content.Server.UserInterface;
+using Content.Shared.DoAfter;
+using Content.Shared.Interaction;
+using Content.Shared.PipePainter;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.PipePainter;
+
+/// <summary>
+/// A system for painting gas pipes using pipe painter
+/// </summary>
+public sealed class PipePainterSystem : EntitySystem
+{
+    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+    [Dependency] private readonly SharedDoAfterSystem _doAfterSystem = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly AtmosPipeColorSystem _pipeColorSystem = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<PipePainterComponent, AfterInteractEvent>(AfterInteractOn);
+        SubscribeLocalEvent<PipePainterComponent, ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<PipePainterComponent, PipePainterSpritePickedMessage>(OnColorPicked);
+        SubscribeLocalEvent<PipePainterComponent, PipePainterDoAfterEvent>(OnDoAfter);
+        SubscribeLocalEvent<PipePainterComponent, ComponentInit>(OnInit);
+    }
+
+    private void OnInit(EntityUid uid, PipePainterComponent component, ComponentInit args)
+    {
+        if (component.ColorPalette.Count == 0)
+            return;
+
+        var startColor = component.ColorPalette.First();
+        SetColor(uid, component, startColor.Key);
+    }
+
+    /// <summary>
+    /// Do after painting action is complete - change pipe color & play sound
+    /// </summary>
+    private void OnDoAfter(EntityUid uid, PipePainterComponent component, PipePainterDoAfterEvent args)
+    {
+        if (args.Handled || args.Cancelled)
+            return;
+
+        args.Handled = true;
+
+        if (args.Args.Target is null)
+            return;
+        var target = (EntityUid) args.Args.Target;
+
+        if(!TryComp<AtmosPipeColorComponent>(target, out var atmosPipeColorComp))
+            return;
+
+        _audio.PlayPvs(component.SpraySound, uid);
+        _pipeColorSystem.SetColor(target, atmosPipeColorComp, args.Color);
+    }
+
+    /// <summary>
+    /// When client picks color. Listens to PipePainterSpritePickedMessage.
+    /// </summary>
+    private void OnColorPicked(EntityUid uid, PipePainterComponent component, PipePainterSpritePickedMessage args)
+    {
+        SetColor(uid, component, args.Key);
+    }
+
+    /// <summary>
+    /// Do on activation - open UI
+    /// </summary>
+    private void OnActivate(EntityUid uid, PipePainterComponent component, ActivateInWorldEvent args)
+    {
+        if (!EntityManager.TryGetComponent(args.User, out ActorComponent? actor))
+            return;
+
+        UpdateUiState(uid, component);
+        component.Owner.GetUIOrNull(PipePainterUiKey.Key)?.Open(actor.PlayerSession);
```

Comment:
```suggestion
        _userInterfaceSystem.TryOpen(uid, PipePainterUiKey.Key, actor.PlayerSession);
```

This should work (I tested).